### DISCUSSION TITLE
Scheduled posts

### DISCUSSION THREAD

**Discussion by AnthonyRonning** [stackernews/stacker.news#3014] [Ideas]
**Is your feature request related to a problem? Please describe.**
I often have my best ideas around midnight, however posting during off-hours will lead to very little engagement. I often sleep in or have other normal life priorities during the more engaging morning hours.

**Describe the solution you'd like**
I would like to draft a post that is scheduled to go out at a specific time in the future. I really like shipyard.pub for scheduling posts on nostr. 

<img width="748" alt="Screenshot 2024-01-09 at 9 26 58 PM" src="https://github.com/stackernews/stacker.news/assets/101225832/d9e88c1a-3f98-4c8b-8366-1a0dbdbf64a4">

**Describe alternatives you've considered**
I could write a bot that does this but would rather not. I could wake up earlier and drink more coffee while using stacker.news but would rather not. 

**Additional context**
Specifically I'm thinking about this for territories since I want to write more engaging posts but it's hard to be on top of it throughout the day each day. I could schedule a bunch of things at once to have content for every day.

**Comment by alexlwn123** [comment]
Strategy/Plan
* mirror `@delete` job flow (`@schedule`)
	* `@schedule in X [seconds|minutes|etc.]`
	* `@schedule for [timestamp]`
	* should work for all kinds of messages (comment, poll, post, etc.)
* Data Model 
	* Add `postScheduledAt` field to Item model
	* Add `isScheduled` field too
	* [link](https://github.com/alexlwn123/stacker.news/blob/d1ed72bb859b11999b76a294c8256ef9967c6acf/api/typeDefs/item.js#L74)
* On message creation (upsert), enque a `ScheduledPostJob`
	* replace `@schedule in X...` with `@schedule for [timestamp]`
	* Only notify the poster (not subscribers)
	*  [link](https://github.com/stackernews/stacker.news/blob/master/api/resolvers/item.js#L1353)
* scheduled posts should be filtered from queries 
	* filter item queries for `!isScheduled`
	* except for on a stacker's own profile page
		* Filter by "scheduled"
* If a scheduled post is deleted
	* delete the item
	* clear the scheduled post job
	~* refund stacker?~
* If a scheduled post is edited
	* clear the schedule post job
	* recreate the schedule job when the stacker hits save
		* Allows stacker to update when the post is scheduled
* When it's time to post the item
    ~* remove `@schedule for ...` line~
    * `updateItem` to `item.isScheduled = false`
    * notify founder,subscribers,poster

**Comment by huumn** [comment]
Excellent breakdown!

IMO:
- No refund
- Keep `@schedule` in the post ... I like the transparency of it.

**Comment by ekzyis** [comment]
Damn, using `@schedule` like `@delete` is a very nice idea. I was worried about having some complicated UI with calendar inputs + time somehow. And relative inputs would still be missing then.

Now this can reuse a lot of stuff from `@delete`!

**Comment by huumn** [comment]
Agreed. *super duper* clever to just do another native bot

**Comment by ekzyis** [comment]
I think however this is implemented, it should also support drafts. A draft is basically a scheduled post with no time yet / scheduled into infinity.

UX might be different though. For example, not sure if `@draft` as an alias for `@schedule` with no options is good UX.

On another note, posting something is the same as scheduling it for right now.

**Comment by h0dlr** [comment]
https://stacker.news/items/302315
https://stacker.news/items/239952

**Comment by 4G0R4** [comment]
any progress on this? it would be really useful

**Comment by ekzyis** [comment]
> any progress on this? it would be really useful

There's an open PR

**Comment by ed-kung** [comment]
I'm looking forward to this one.  I've wanted to use it a lot, for a lot of different purposes.

**Comment by pory-gone** [comment]
Hello, 
is anyone currently working on implementing this?
With the new Lexical editor, i was thinking about something. Implementing `@schedule` commands, that behave like `@delete` and `@remindme`, would be a straightforward solution, but Lexical opens up the possibility of adding a scheduling button directly in the toolbar.

Would a hybrid implementation be appreciated? 

Happy to hear opinions and help!

**Comment by huumn** [comment]
This should work like `@remindme` but it's WAY more complicated than that. Whether you use lexical or not is dessert.  

If you'd like to work on this, come up with a plan/design of how it will work first.
